crates/wit-parser/src/ast/expand.rs

```diff
@@ -0,0 +1,100 @@
+use crate::{Resolve, World, WorldId, PackageId, WorldItem, InterfaceId};
+
+pub type Substitutions = HashMap<String, HashMap<String, HashSet<String>>>;
+
+pub fn expand(
+    resolve: &mut Resolve,
+    package_id: PackageId,
+    world: Option<&str>,
+    substitutions: Substitutions
+) -> Result<()> {
+    let world_id = resolve.select_world(package_id, world)?;
+    resolve.expand_wildcards(world_id, substitutions)
+}
+
+impl Resolve {
+    pub fn expand_wildcards(&mut self, world_id: WorldId, substitutions: Substitutions) -> Result<()> {
+        let mut world = self.worlds.get(world_id).cloned().ok_or(anyhow!("invalid world id"))?;
+        self.expand_world(&mut world, substitutions)?;
+        self.worlds[world_id] = world;
+        Ok(())
+    }
+
+    fn has_wildcards(&self, interface_id: InterfaceId) -> bool {
+        self
+            .interfaces[interface_id]
+            .wildcard
+            .is_some()
+    }
+
+    fn expand_world(&mut self, world: &mut World, mut substitutions: Substitutions) -> Result<()> {
+        let mut world_subs = substitutions.remove(&world.name).unwrap_or_default();
+
+        self.expand_imports(world, &mut world_subs)?;
+        self.expand_exports(world, &mut world_subs)?;
+
+        if !world_subs.is_empty() {
+            log::warn!(
+                "unused substitutions were provided for expanding world {name:?}",
+                name = world.name,
+            );
+        }
+
+        Ok(())
+    }
+
+    fn expand_imports(
+        &mut self,
+        world: &mut World,
+        substitutions: &mut HashMap<String, HashSet<String>>
+    ) -> Result<()> {
+        self.expand_items(&world.name, "import", &mut world.imports, substitutions)
+    }
+
+    fn expand_exports(
+        &mut self,
+        world: &mut World,
+        substitutions: &mut HashMap<String, HashSet<String>>
+    ) -> Result<()> {
+        self.expand_items(&world.name, "export", &mut world.imports, substitutions)
+    }
+
+    fn expand_items<'a>(
+        &mut self,
+        world_name: &str,
+        desc: &str,
+        items: &mut IndexMap<String, WorldItem>,
+        substitutions: &mut HashMap<String, HashSet<String>>
+    ) -> Result<()> {
+        for (name, item) in items {
+            if let WorldItem::Interface(interface_id) = item {
+                if self.has_wildcards(*interface_id) {
+                    let iface_subs = substitutions
+                        .remove(name)
+                        .ok_or_else(|| anyhow!(
+                            "world {world_name} {desc} item {name} contains wildcards but no substitutions were provided",
+                        ))?;
+                    *interface_id = self.expand_interface(*interface_id, iface_subs); 
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn expand_interface(&mut self, interface_id: InterfaceId, subnames: HashSet<String>) -> InterfaceId {
+        let mut new_interface = self.interfaces[interface_id].clone();
```

Comment:
Ah unfortunately cloning this won't be so simple I think due to the fact that there's lots of "back references" to the owner of an item. For example types have an `InterfaceId` as their owner and that would also need to be updated here. There's additionally references from the document to its interfaces that it contains.

Overall I'm not sure though that cloning and expanding like this will work out. Would it be possible to fill-out in-place? For example basically converting a wildcard-containing interface to one without wildcards?

Reply:
Ahh you're absolutely correct. I'll have to give this more thought.

Funny enough, my original approach was expanding in-place...However, templated interfaces need to be expanded more-than-once which is the motivation for alloc'ing new interfaces.

I'm guessing it would be quite painful to patch the back references to expanded interfaces in types and documents.

Reply:
After chatting with @dicej, a viable approach might be to reconstruct `Resolve` as we are expanding instead of trying to mutably update as is done now. I anticipate the tricky part of this will be how to reconcile the topological order of things while inserting new interfaces.

Reply:
That might work, yeah, but personally I don't really know a good way of handling all this. You can take a look at the support in `resolve.rs` for merging two `Resolve`s to see what the order-of-magnitude is for ids that need remapping. 

One other tricky part is that if a wildcard interface is both exported and imported then the import/export refer to the same ID, but if you can specify different substitutions for both then that ID mapping needs to become different. Then that introduces two interfaces with possibly clashing names which also further needs resolution.